php/libraries/DefaultFilesUploadHandler.class.inc

```diff
@@ -0,0 +1,367 @@
+     * file uploads have a value (i.e. no "optional" upload is allowed). This
+     * behavior can be overridden bu subclasses should they need to.
+     *
+     * @return bool true if the uploaded file set is valid, false otherwise.
+     */
+    public function isValid() : bool
+    {
+        $valid = true;
+        foreach ($this->getFiles() as $field => $file) {
+            if ($file === false || $file['error'] != UPLOAD_ERR_OK) {
+                $this->addError(
+                    $field,
+                    $file === false ? false : $file['error']
+                );
+                $valid = false;
+            }
+        }
+
+        return $valid;
+    }
+
+    /**
+     * Main method for this handler. Validates the uploaded files (i.e. make sure
+     * that they are are present on the server) and moves them to their final
+     * locations.
+     *
+     * @return nothing.
+     */
+    public function processAll()
+    {
+        try {
+            if ($this->isValid()) {
+                foreach ($this->getFiles() as $field => $file) {
+                    if ($file !== false) {
+                        $newFile = $this->getNewLocation($file['tmp_name']);
+                        $newDir  = $newFile->getPath();
+
+                        // Check to see if directory exists
+                        if (!file_exists($newDir)) {
+                            // Directory does not exist: try to create it
+                            if (!mkdir($newDir, $this->_dirPermissions, true)) {
+                                $this->addError(
+                                    $field,
+                                    "Cannot move {$file['tmp_name']} to $newDir:"
+                                        . " directory $newDir could not be created"
+                                );
+                            } else {
+                                if ($this->_dirGroup != '') {
+                                    if (!chgrp($newDir, $this->_dirGroup)) {
+                                        $this->addError(
+                                            $field,
+                                            "Command 'chgrp {$this->_dirGroup} "
+                                                . "$newDir' failed"
+                                        );
+                                    }
+                                }
+                            }
+                            // File name exists. Make sure it refers to a directory.
+                        } else if (!is_dir($newDir)) {
+                            $this->addError(
+                                $field,
+                                "Cannot move {$file['tmp_name']} to $newDir: "
+                                    . "$newDir already exists and is not a directory"
+                            );
+                        }
+
+                        // If everything went OK so far
+                        if (!isset($this->_errors[$field])) {
```

Comment:
This if statement could be skipped if a `continue` was added after the `addError` call